### PR TITLE
feat(node): Add api endpoints to detect/install node

### PR DESCRIPTION
**Pull request by mrmeku**

### CODE REVIEW

File: apps/angular-console/src/app/app.module.ts [hunk 1]
```diff
@@ -88,7 +89,12 @@ export function initApollo(
         {
           path: '',
           children: workspaceRoutes,
-          canActivateChild: [CancelCommandGuard]
+          canActivateChild: [IsNodeJsInstalledGuard, CancelCommandGuard]
+        },
+        {
+          path: 'install-nodejs',
+          loadChildren:
+            '@angular-console/feature-install-node-js#FeatureInstallNodeJsModule'
```
Comment (vsavkin) [inline apps/angular-console/src/app/app.module.ts]: This is funky looking URL!!

Comment (vsavkin) [reply apps/angular-console/src/app/app.module.ts]: Actually, it is pretty good

Comment (mrmeku) [reply apps/angular-console/src/app/app.module.ts]: :p, don't blame me, blame the author of @angular/router

File: apps/angular-console/src/app/app.module.ts [hunk 2]
```diff
@@ -88,7 +89,12 @@ export function initApollo(
...
         },
         {
           path: 'support',
```
Comment (vsavkin) [inline apps/angular-console/src/app/app.module.ts]: We should probably make support lazy as well in a later PR

File: libs/feature-install-node-js/src/lib/feature-install-node-js.module.spec.ts [hunk 3]
```diff
@@ -0,0 +1,14 @@
+import { async, TestBed } from '@angular/core/testing';
```
Comment (vsavkin) [inline libs/feature-install-node-js/src/lib/feature-install-node-js.module.spec.ts]: Delete an empty test

File: libs/feature-install-node-js/src/lib/install-node-js.component.spec.ts [hunk 4]
```diff
@@ -0,0 +1,24 @@
+import { async, ComponentFixture, TestBed } from '@angular/core/testing';
```
Comment (vsavkin) [inline libs/feature-install-node-js/src/lib/install-node-js.component.spec.ts]: Delete an empty spec

File: libs/feature-install-node-js/src/lib/install-node-js.component.ts [hunk 5]
```diff
@@ -0,0 +1,71 @@
+import { Component } from '@angular/core';
+import { Router } from '@angular/router';
+import { Apollo } from 'apollo-angular';
+import gql from 'graphql-tag';
+import { Observable } from 'rxjs';
+import { map, switchMap, tap } from 'rxjs/operators';
+
+import { InstallNodeJsStatus } from '../../../../server/src/graphql-types';
```
Comment (vsavkin) [inline libs/feature-install-node-js/src/lib/install-node-js.component.ts]: could you remove this and replace this with any for now. I don't think the client-side code should reference the server folder.

Once nx 6.4 is out, I'm going to upgrade console to use it, so we can share libs between the server and the client. Then we can share all the types.

File: libs/feature-install-node-js/src/lib/install-node-js.component.ts [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+                }
+              `
+            })
+            .valueChanges.pipe(
+              map(
+                (result: {
+                  data: { installNodeJsStatus: InstallNodeJsStatus };
+                }) => result.data.installNodeJsStatus
+              ),
+              tap(res => console.log(res))
```
Comment (vsavkin) [inline libs/feature-install-node-js/src/lib/install-node-js.component.ts]: this should not be here

File: libs/feature-install-node-js/src/lib/install-node-js.component.ts [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+            ) as Observable<InstallNodeJsStatus>;
+        }),
+        tap(({ success }: InstallNodeJsStatus) => {
+          if (success) {
```
Comment (vsavkin) [inline libs/feature-install-node-js/src/lib/install-node-js.component.ts]: I think we should handle error as well.

You can still redirect to console but you can also use messanger service to show some error message.

File: libs/utils/src/lib/nodejs-install.service.ts [hunk 8]
```diff
@@ -0,0 +1,37 @@
+import { Injectable } from '@angular/core';
+import gql from 'graphql-tag';
+
+import { CommandRunner } from './command-runner.service';
+
+@Injectable({ providedIn: 'root' })
+export class NodejsInstall {
```
Comment (vsavkin) [inline libs/utils/src/lib/nodejs-install.service.ts]: This service isn't used.

File: server/src/api/install-nodejs.ts [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+  nodeDownload = null;
+  nodeInstallDone = false;
+  try {
+    switch (platform()) {
+      case 'darwin':
+        nodeDownloadProgress = progress({
+          length: 15546624,
+          time: 100
+        }) as Progress;
+        const pkg = fs.createWriteStream('node.pkg').on('finish', () => {
```
Comment (vsavkin) [inline server/src/api/install-nodejs.ts]: you should use the tmp directory. you don't want to write it into the current directory

Comment (vsavkin) [reply server/src/api/install-nodejs.ts]: You can use `import { dirSync } from 'tmp';`

File: apps/angular-console/src/app/app.module.ts (hunk 1 above)
Comment (vsavkin) [reply apps/angular-console/src/app/app.module.ts]: :) I was just testing out hte new vscode github integration

https://code.visualstudio.com/blogs/2018/09/10/introducing-github-pullrequests

Turns out you can add comments, but you cannot delete them :)

**Comment by vsavkin** [comment]
Delete node.pkg file

**Comment by vsavkin** [comment]
Looks very good. Great job, Dan! I left a few minor comments.

**Comment by vsavkin** [comment]
Just tried it on Windows and I'm seeing the following issue:

The frontend keeps receiving these:

```
{downloadPercentage: 100, downloadSpeed: 114403.06086956522, success: null, cancelled: null, error: null, …}
```

I imagine is because this line `exec(`msiexec /i "${msi.path}"`);` needs to listen to `on('close')` to set nodeInstallDone to true.

File: apps/angular-console/src/app/app.module.ts (hunk 2 above)
Comment (mrmeku) [reply apps/angular-console/src/app/app.module.ts]: Agreed, should speed up boot up time a bit more (not that Jason hasn't already solved that)

File: libs/feature-install-node-js/src/lib/install-node-js.component.ts (hunk 5 above)
Comment (mrmeku) [reply libs/feature-install-node-js/src/lib/install-node-js.component.ts]: Donezo

File: libs/feature-install-node-js/src/lib/install-node-js.component.ts (hunk 6 above)
Comment (mrmeku) [reply libs/feature-install-node-js/src/lib/install-node-js.component.ts]: donzo

File: libs/utils/src/lib/nodejs-install.service.ts (hunk 8 above)
Comment (mrmeku) [reply libs/utils/src/lib/nodejs-install.service.ts]: Donezo

**Comment by vsavkin** [comment]
Merged here:
https://github.com/nrwl/angular-console/pull/242

Made two changes:
- Fixed the window version (msi.path is absolute)
- I also added an option to skip the install. This is in case they don't have node in path, which is not a completely unthinkable thing.

Cheers!

**Pull request closed without merging**